src/csharp/Grpc.Core/Internal/BinaryStringDictionary.cs

```diff
@@ -0,0 +1,130 @@
+        }
+
+        public void Add(string str)
+        {
+            dict.Add(new BinaryKey(EncodingACII.GetBytes(str)), str);
+        }
+
+        /// <summary>
+        /// Looks up the string based on its ASCII-encoded binary form.
+        /// </summary>
+        public string Lookup(IntPtr nativeData, int nativeDataLen)
+        {
+            var key = new BinaryKey(nativeData, nativeDataLen);
+            dict.TryGetValue(key, out string str);
+            return str;
+        }
+        
+        private struct BinaryKey : IEquatable<BinaryKey>
+        {
+            private byte[] data;
+            private IntPtr nativeData;
+            private int nativeDataLen;
+
+            public BinaryKey(byte[] data)
+            {
+                this.data = GrpcPreconditions.CheckNotNull(data);
+                this.nativeData = IntPtr.Zero;
+                this.nativeDataLen = 0;
+            }
+
+            public BinaryKey(IntPtr nativeData, int nativeDataLen)
+            {
+                GrpcPreconditions.CheckArgument(nativeData != IntPtr.Zero);
+                GrpcPreconditions.CheckArgument(nativeDataLen >= 0);
+                this.data = null;
+                this.nativeData = nativeData;
+                this.nativeDataLen = nativeDataLen;
+            }
+
+            /// <summary>
+            /// Indicates whether this instance and a specified object are equal.
+            /// </summary>
+            public override bool Equals(object obj)
+            {
+                return obj is BinaryKey && Equals((BinaryKey)obj);
+            }
+
+            /// <summary>
+            /// Returns the hash code for this instance.
+            /// </summary>
+            public override int GetHashCode()
+            {
+                return ComputeHashCode(GetSpan());
+            }
+
+            /// <summary>
+            /// Indicates whether this instance and a specified object are equal.
+            /// </summary>
+            public bool Equals(BinaryKey other)
+            {
+                return GetSpan().SequenceEqual(other.GetSpan());
+            }
+
+            private ReadOnlySpan<byte> GetSpan()
+            {
+                if (this.data != null)
+                {
+                    return new ReadOnlySpan<byte>(this.data);
+                }
+                if (this.nativeData != IntPtr.Zero)
+                {
+                    unsafe { return new ReadOnlySpan<byte>((byte*)this.nativeData, this.nativeDataLen); }
+                }
+                throw new ArgumentNullException("Content cannot be null.");
+            }
+
+            private static int ComputeHashCode(ReadOnlySpan<byte> data)
+            {
+                unchecked
+                {
+                    // TODO(jtattermusch): there should be a faster way of computing some meaninful hashcode
+                    // for a block of memory, but for now this should be sufficient.
+                    int hc = 1768953197;
```

Comment:
just to double check: this hash code is using going to experience int overflow while it's being computed, which is always ok (e.g. defined behavior)?